include/mbedtls/ssl_internal.h

```diff
@@ -290,6 +292,86 @@ struct mbedtls_ssl_handshake_params
                     const unsigned char *, size_t,
                     unsigned char *, size_t);
 
+    /*
+     * State-local variables used during the processing
+     * of a specific handshake state.
+     *
+     */
+
+    union
```

Comment:
I'm not really convinced to this solution. It seems like a mimicry of a polymorphic structure.
I understand that you're trying to conserve memory, but this structure does not seem to be in quantities in which an optimization like this would save significant memory ammounts.
This union also does not seem to be useful without an external variable indicating its state, and if someone reads its contents between changing the external variable and the state of an union, undefined behaviour occurs.
For most of the structs only a "preparation_done" is needed, and only three out of eight add other fields - so maybe state-specific structs should be used, and a state indicator, which would substitute all of the preparation_done fields?
I'll look into it further after a first round of a general review.

Reply:
@AndrzejKurek The union is indexed by the pre-existing handshake state variable `mbedtls_ssl_context::state`. 

I agree that currently the savings of using a union are not significant, but I don't see a reason not to use a union nonetheless. If not more, it helps to catch errors when using these variables outside there intended scope. 

@AndrzejKurek @mpg Note: If, in DTLS, we want to re-write handshake messages on retransmitting instead of buffering them, this union will need to be split on a per-flight basis, not on a per-message basis.

Reply:
@AndrzejKurek Regarding the extraction of `preparation_done`, I agree that it might be worth having an additional field indicating the state of processing, independent of the field(s) used to hold temporary data used during the processing. Let's wait for @mpg's opinion and potentially make the change then.